examples/main.rs

```diff
@@ -16,57 +20,128 @@ fn main() {
     println!("ColorType is {:?}", rgb_img.color());
 
     let mut rgba_img = ImageBuffer::new(width, height);
+    let mut rg_img = ImageBuffer::new(width, height);
+    let mut r_img = ImageBuffer::new(width, height);
 
-    println!("Converting RGB -> RGBA"); // could be optimized
+    println!("Converting RGB -> RGBA/RG/R"); // could be optimized
     for x in 0u32..width {
         for y in 0u32..height {
             let pixel = rgb_img.get_pixel(x, y);
             let pixel_rgba = pixel.to_rgba();
+            let pixel_rg = LumaA::from([pixel_rgba[0], pixel_rgba[1]]);
+            let pixel_r = Luma::from([pixel_rgba[0]]);
             rgba_img.put_pixel(x, y, pixel_rgba);
+            rg_img.put_pixel(x, y, pixel_rg);
+            r_img.put_pixel(x, y, pixel_r);
         }
     }
 
     let block_count = intel_tex_2::divide_up_by_multiple(width * height, 16);
     println!("Block count: {}", block_count);
-
-    let mip_count = 1;
-    let array_layers = 1;
-    let caps2 = Caps2::empty();
-    let is_cubemap = false;
-    let resource_dimension = D3D10ResourceDimension::Texture2D;
-    let alpha_mode = AlphaMode::Opaque;
-    let depth = 1;
-
-    let mut dds = Dds::new_dxgi(NewDxgiParams {
+    let dds_defaults = NewDxgiParams {
         height,
         width,
-        depth: Some(depth),
-        format: DxgiFormat::BC7_UNorm,
-        mipmap_levels: Some(mip_count),
-        array_layers: Some(array_layers),
-        caps2: Some(caps2),
-        is_cubemap,
-        resource_dimension,
-        alpha_mode,
-    })
-    .unwrap();
-
-    let surface = intel_tex_2::RgbaSurface {
-        width,
-        height,
-        stride: width * 4,
-        data: &rgba_img,
+        depth: Some(1),
+        format: DxgiFormat::Unknown,
+        mipmap_levels: Some(1),
+        array_layers: Some(1),
+        caps2: Some(Caps2::empty()),
+        is_cubemap: false,
+        resource_dimension: D3D10ResourceDimension::Texture2D,
+        alpha_mode: AlphaMode::Opaque,
     };
+    // BC4
+    {
+        let mut dds = Dds::new_dxgi(NewDxgiParams {
+            format: DxgiFormat::BC4_UNorm,
+            ..dds_defaults
+        })
+        .unwrap();
+
+        let stride = width;
+        assert_eq!(
+            dds.get_pitch(),
+            Some(stride * 2), //the output is 2 bytes per pixel
+            "dds stride ({:?}) did not match the provided surface stride ({})",
+            dds.get_pitch(),
+            stride * 2
+        );
 
-    println!("Compressing to BC7...");
-    bc7::compress_blocks_into(
-        &bc7::opaque_ultra_fast_settings(),
-        &surface,
-        dds.get_mut_data(0 /* layer */).unwrap(),
-    );
-    println!("  Done!");
+        let surface = intel_tex_2::RSurface {
+            width,
+            height,
+            stride,
```

Comment:
Shouldn't this be `*2`?

Idea was to have `let stride = width * 2` :)

Reply:
In hindsight I think none of what I was saying makes sense. The `stride` in `intel_tex_2::*Surface` describes the _input stride_. While the DDS file and the raw BCn blob describes the _output_ stride. This crate doesn't have an explicit way to converse about the latter, it's "just a mutable slice to write to" and not much else.